<a href="https://colab.research.google.com/github/marce3-2140/Unsupervised-ML-Models/blob/main/PCA_LinearRegressor.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [15]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.decomposition import PCA
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [1]:
#load the dataset
from sklearn.datasets import fetch_openml
mnist = fetch_openml('mnist_784')

mnist.data.shape

/usr/local/lib/python3.10/dist-packages/sklearn/datasets/_openml.py:968: FutureWarning: The default value of `parser` will change from `'liac-arff'` to `'auto'` in 1.4. You can set `parser='auto'` to silence this warning. Therefore, an `ImportError` will be raised from 1.4 if the dataset is dense and pandas is not installed. Note that the pandas parser may return different data types. See the Notes Section in fetch_openml's API doc for details.
  warn(


(70000, 784)

In [5]:
scaler = StandardScaler()
scaled_df = scaler.fit_transform(mnist.data)

In [6]:
pca = PCA()
pca.fit(scaled_df)

PCA()

In [10]:
%%time
pca = PCA(n_components=0.95)
pca_data = pca.fit_transform(scaled_df)

print("El número de componentes es", pca.n_components_, "de", pca.n_features_)

El número de componentes es 332 de 784
CPU times: user 24.4 s, sys: 1.44 s, total: 25.8 s
Wall time: 15.3 s


/usr/local/lib/python3.10/dist-packages/sklearn/utils/deprecation.py:101: FutureWarning: Attribute `n_features_` was deprecated in version 1.2 and will be removed in 1.4. Use `n_features_in_` instead.
  warnings.warn(msg, category=FutureWarning)


In [11]:
# Dividir los datos en conjuntos de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(pca_data, mnist.target, test_size=0.2, random_state=42)

In [21]:
# Entrenar un modelo utilizando los datos transformados en PCA
%%time
model_with_pca = LogisticRegression(max_iter=1000)
model_with_pca.fit(X_train, y_train)

CPU times: user 3min 29s, sys: 21.5 s, total: 3min 51s
Wall time: 2min 12s


LogisticRegression(max_iter=1000)

In [22]:
# Realizar predicciones en el conjunto de prueba
y_pred_with_pca = model_with_pca.predict(X_test)

In [23]:
# Calcular la precisión del modelo
accuracy_with_pca = accuracy_score(y_test, y_pred_with_pca)
print("Precisión del modelo con PCA:", accuracy_with_pca)

Precisión del modelo con PCA: 0.922


In [24]:
# También, puedes entrenar un modelo utilizando los datos originales sin PCA (pero escalados)
X_train_orig, X_test_orig, _, _ = train_test_split(scaled_df, mnist.target, test_size=0.2, random_state=42)

In [25]:
%%time
model_without_pca = LogisticRegression(max_iter=1000)
model_without_pca.fit(X_train_orig, y_train)

CPU times: user 8min 45s, sys: 30.1 s, total: 9min 15s
Wall time: 5min 17s


LogisticRegression(max_iter=1000)

In [26]:
y_pred_without_pca = model_without_pca.predict(X_test_orig)
accuracy_without_pca = accuracy_score(y_test, y_pred_without_pca)
print("Precisión del modelo sin PCA:", accuracy_without_pca)

Precisión del modelo sin PCA: 0.9165714285714286


> **Conclusiones**

  * el "Modelo con PCA" tiene una precisión ligeramente mejor (92%) en comparación con el "Modelo sin PCA" (91%). Esto sugiere que la reducción de dimensionalidad con PCA ha mejorado el rendimiento del modelo en este conjunto de datos.
  * Es importante destacar que incluso una pequeña mejora en la precisión puede ser significativa en algunos casos, especialmente si se trata de una tarea crítica. Por lo tanto, en esta situación, es recomendable utilizar el "Modelo con PCA" debido a su mejor rendimiento.